rust/nasl-interpreter/src/built_in_functions/cryptography/aes_ccm.rs

```diff
@@ -0,0 +1,270 @@
+// Copyright (C) 2023 Greenbone Networks GmbH
+//
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+use ::aes::{Aes128, Aes192, Aes256};
+use aes::cipher::{BlockCipher, BlockDecrypt, BlockEncrypt, BlockSizeUser};
+use ccm::{
+    aead::{Aead, Error as aError},
+    consts::{U10, U11, U12, U13, U16, U7, U8, U9},
+    Ccm, KeyInit, NonceSize, TagSize,
+};
+use digest::generic_array::ArrayLength;
+use sink::Sink;
+
+use crate::{error::FunctionError, NaslFunction, NaslValue, Register};
+
+use super::{get_named, Crypt};
+
+/// Function to create cipher object and en-/decrypt data. Can throw error in case of authentication failure.
+fn ccm_iv_len<D, M, N>(
+    crypt: Crypt,
+    key: &[u8],
+    nonce: &[u8],
+    data: &[u8],
+) -> Result<Vec<u8>, aError>
+where
+    D: BlockCipher + BlockSizeUser<BlockSize = U16> + BlockEncrypt + BlockDecrypt + KeyInit,
+    M: ArrayLength<u8> + TagSize,
+    N: ArrayLength<u8> + NonceSize,
+{
+    let cipher = Ccm::<D, M, N>::new(key.into());
+    match crypt {
+        Crypt::Encrypt => cipher.encrypt(nonce.into(), data),
+        Crypt::Decrypt => cipher.decrypt(nonce.into(), data),
+    }
+}
+
+/// Base function for ccm en- and decryption. Sets the tag length to 16.
+fn ccm<D>(register: &Register, crypt: Crypt, function: &str) -> Result<NaslValue, FunctionError>
+where
+    D: BlockCipher + BlockSizeUser<BlockSize = U16> + BlockEncrypt + BlockDecrypt + KeyInit,
+{
+    // Get parameters
+    let key = get_named(register, "key", true, function)?.unwrap();
+    let data = get_named(register, "data", true, function)?.unwrap();
+    let nonce = get_named(register, "iv", true, function)?.unwrap();
+    // Switch mode dependent on iv length
+    let res = match nonce.len() {
+        7 => ccm_iv_len::<D, U16, U7>(crypt, key, nonce, data),
+        8 => ccm_iv_len::<D, U16, U8>(crypt, key, nonce, data),
+        9 => ccm_iv_len::<D, U16, U9>(crypt, key, nonce, data),
+        10 => ccm_iv_len::<D, U16, U10>(crypt, key, nonce, data),
+        11 => ccm_iv_len::<D, U16, U11>(crypt, key, nonce, data),
+        12 => ccm_iv_len::<D, U16, U12>(crypt, key, nonce, data),
+        13 => ccm_iv_len::<D, U16, U13>(crypt, key, nonce, data),
```

Comment:
This part is planed to be moved into a macro and also support setting the Tag length. Currently the Tag length is hardcoded to 16 Bytes (which was also the case for NASL c: https://github.com/greenbone/openvas-scanner/blob/main/nasl/nasl_crypto2.c#L1837)